pkg/internal/cluster/create/actions/config/config.go

```diff
@@ -121,22 +134,21 @@ func (a *Action) Execute(ctx *actions.ActionContext) error {
 }
 
 // getKubeadmConfig generates the kubeadm config contents for the cluster
-// by running data through the template.
-func getKubeadmConfig(cfg *config.Cluster, data kubeadm.ConfigData) (path string, err error) {
+// by running data through the template and applying patches as needed.
+func getKubeadmConfig(data kubeadm.ConfigData, patches []string, jsonPatches []config.PatchJSON6902) (path string, err error) {
 	// generate the config contents
-	config, err := kubeadm.Config(data)
+	cf, err := kubeadm.Config(data)
 	if err != nil {
 		return "", err
 	}
-	// fix all the patches to have name metadata matching the generated config
-	patches, jsonPatches := allPatchesFromConfig(cfg)
+
 	// apply patches
-	// TODO(bentheelder): this does not respect per node patches at all
-	// either make patches cluster wide, or change this
-	patched, err := patch.Patch(config, patches, jsonPatches)
+	patched, err := patch.Patch(cf, patches, jsonPatches)
```

Comment:
This is going to have surprising behavior in the current implementation. All merge patches will be applied and then all 6902 patches.

As a user I would instead expect all cluster global patches and then all node level patches

Reply:
Yes, I agree as a user I would expect the same.  I didn't think about it throughly. So, something like this maybe? Of course, I would need to provide patches separately and better code. wdyt?

```Go
clusterPatched, err := patch.Patch(cf, clusterPatches, clusterJSONPatches)
...
allPatched, err := patch.Patch(clusterPatched nodePatches, nodeJSONPatches)
...

```

Reply:
I think this works, I would add that we can skip running the second patch logic in the case that there are no patches, but we should always run the first patch (it also has the behavior of normalizing the output).

```go
patchedConfig, err := patch.Patch(cf, clusterPatches, clusterJSONPatches)
// handle error
if len(nodePatches) > 0 || len(nodeJSONPatches) > 0 {
  patchedConfig, err = patch.Patch(patchedConfig, nodePatches, nodeJSONPatches)
  // handle error
}
```

Reply:
@BenTheElder  I ran into an interesting situation - **config.Node** vs **nodes.Node**.  As you know, the later is what I have access to within the **getKubeadmConfig** function, so how do I match the right **config.Node.'KubeadmConfigPatches'** patch(es) to apply it to the current  **nodes.Node** without having anything to correlate with?  I thought about the Role, but that's is not specific enough. The array indexes from nodes.Node(i) == cfg.Nodes[i] might do it, but is the index reliable in this case?  thoughts?